### PR TITLE
Don't use process dictionary for bump_reduce_memory_use message

### PR DESCRIPTION
**Pull request by lukebakken**
Alternative for #1388 that does not use process dictionary.

### CODE REVIEW

**Comment by hairyhum** [comment]
This will change the `rabbit_backing_queue` behaviour and cannot go to a patch release. That's why I used the process dictionary.

**Comment by lukebakken** [comment]
OK I'll re-target it for `master` once `rabbitmq-server-batch-betas` is merged 😄

File: src/rabbit_amqqueue_process.erl [hunk 1]
```diff
@@ -1454,10 +1454,10 @@ handle_info({bump_credit, Msg}, State = #q{backing_queue       = BQ,
...
     credit_flow:handle_bump_msg(Msg),
     noreply(State#q{backing_queue_state = BQ:resume(BQS)});
-handle_info(bump_reduce_memory_use, State = #q{backing_queue       = BQ,
-                                               backing_queue_state = BQS}) ->
-    put(waiting_bump, false),
-    noreply(State#q{backing_queue_state = BQ:resume(BQS)});
+handle_info(bump_reduce_memory_use, State = #q{ backing_queue       = BQ,
+                                                backing_queue_state = BQS0 }) ->
+    BQS1 = BQ:handled_bump_reduce_memory_use(BQS0),
+    noreply(State#q{ backing_queue_state = BQ:resume(BQS1) });
```
Comment (hairyhum) [inline src/rabbit_amqqueue_process.erl]: This code calls two callbacks, one of them just updates the state. Should it be a single callback instead?

**Comment by lukebakken** [comment]
@hairyhum thanks for noticing that the new callback is not necessary.

**Comment by hairyhum** [comment]
The callback is actually necessary, because not every `resume` means `received bump`, so there can be another callback, which will set the flag and do `resume`. 
The flag here makes sure that there is only one bump message in the message box, so the message box is not polluted. Every time queue process handles the bump message it should `resume`, but not every time it resumes it should clear the flag.
If there is another way to make sure we have a single bump message in the message box at a time, it can be used instead.

**Comment by lukebakken** [comment]
> Every time queue process handles the bump message it should resume, but not every time it resumes it should clear the flag.

@hairyhum got it. Thanks again for the careful reviews.

**Pull request merged**